test/gie/4D-API_cs2cs-style.gie

```diff
@@ -0,0 +1,89 @@
+===============================================================================
+
+Test the 4D API handling of cs2cs style transformation options.
+
+These tests are mostly based on the same material as those in
+more_builtins.gie, since we are testing the same kinds of things,
+but provided through a different interface.
+
```

Comment:
This should also contain a bunch of tests using the epsg init-file. Test data is easy to prepare with cs2cs. I can come up with some suggestions if needed.

Reply:
That would be nice. What is really needed (and will probably cause hiccups) is cases where more than one of the cs2cs style options are used.

Reply:
I'll merge #705 s soon as all the tests pass. Sync your branch with master and you'll have a bunch of test material in `test/gigs/`. I think that should be enough for this purpose.

Reply:
Have sync'ed, but currently drowning in all the "almost OK" tests. I think there is a bug in the interpretation of the GIGS tolerance parameter.

If I understand @micahcochran's comments correctly, they are considered to be max absolute coordinatewise deviations (i.e. in decimal degrees for latitude/longitude pairs), whereas in the `.gie` version, they are translated into "same numerical value but interpreted as having a *meter* unit".

This makes a.o. all NADCON tests fail, because 1e-7 m (0.0001 mm) is much less than 1e-7 degree (1 m at the Equator), at least if you are more than a few minutes of arc away from the poles.

Reply:
Which tests fail? There's a few of them that doesn't work in the current python setup as well. It might be that it is only those tests that fail, in which case you can disregard them, since the problems are not related to the work you are doing here.

Reply:
I can see from the notes of @micahcochran that a large number is supposed to fail slightly and a much smaller number is supposed to fail miserable.

But getting the general view of more than 600 failing tests takes some time, so I'm starting out in a horrible way by introducing the infinity norm for any tests where `gigs` is part of the path to the test file.

Whatever remains, I'll check case-by-case - or, if many remain, by whatever works.

Reply:
Here https://github.com/OSGeo/proj.4/blob/master/travis/install.sh#L98-L99 you can see which tests are working. Tests not in those two calls will fail. Maybe that helps you a bit?

Reply:
Certainly - down to 223 failures :-) I'll go on debugging tomorrow

Reply:
Happy hunting :)

Reply:
> Happy hunting :)

Already down to 199...

Reply:
Still some problems regarding the handling of torad/todeg in gie, when axis!=enu.

The same problem is, as far as I can see, present in cs2cs: expecting enu, converting first two to radians on input, not taking care of different axis order on neither input, nor output.

Reply:
but repaired the double superfluous datum shifts in 5103.1

Reply:
> but repaired the double superfluous datum shifts in 5103.1

It seems to me that you've just changed the test. Is that correct?

Reply:
> Still some problems regarding the handling of torad/todeg in gie, when axis != enu.

> The same problem is, as far as I can see, present in cs2cs: expecting enu, converting first two fields to radians on input, not taking care of different axis order on neither input, nor output.

When I tried to fix the axis ordering business, I missed your `4D-API_cs2cs-style.gie` test since it wasn't added to the CMake test-suite. I see now that there's a few errors. I'm looking into that now.

Reply:
> It seems to me that you've just changed the test. Is that correct?

I overwrite the `towgs84=...` elements from the epsg file with a null operation whic is, in turn, ignored by the prepare/finalize code.

This makes pj_fwd/pj_inv work the same way as pj_transform: omitting dual identical helmert shifts, so it doesn't really change the test - it rewrites the test in order to act identically to the json case.

The alternative to declaring this *an idiom for signalling "no datum shift here please"*, is to write a large amount of quite intrusive, and probably quite brittle, code in `pj_pipeline.c` doing hand held manipulation across otherwise individual pipeline steps.

I intend to do further clean up of the gigs handling in gie in time for 5.1.0, so this is a simple, non-intrusive stop gap for now.

Reply:
> I'm looking into that now.

In that context, please note that I have merged your previous PR over at busstoptaktik/PROJ.4.

Reply:
> This makes pj_fwd/pj_inv work the same way as pj_transform: omitting dual identical helmert shifts, so it doesn't really change the test - it rewrites the test in order to act identically to the json case.

I understand why you've done it, I just don't think it is going to work. The end goal of this PR is, amongst other things, to be able to put to `epsg` CRS definitions into `proj_create_crs_to_crs()` and get correct results when using the resulting `PJ` in transformations. While your solution fixes the test, it doesn't fix the transformation in the test when used in the wild (in the 4D API).

Reply:
> In that context, please note that I have merged your previous PR over at busstoptaktik/PROJ.4.

Duly noted.

Reply:
> The end goal of this PR is, amongst other things, to be able to put to epsg CRS definitions into `proj_create_crs_to_crs()` and get correct results when using the resulting PJ in transformations

`proj_create_crs_to_crs()` is a much different situation:

**First**, we do not necessarily need this, if we can live with the time and slight precision penalty:  For typical crs_to_crs uses we do just one half-roundtrip, and do not accumulate helmert imprecision.

**Second**, we can implement the `pj_transform` trick as an optimization in a much simpler way than in the generic `pipeline` case: Here we know we have a one input, one output situation, and if both are from epsg definitions we can check whether the `towgs84=...` incantations are identical and bypass Helmert by setting the `towgs84=0.0.0` thingy at the `proj=pipeline` level.

This can be economically implemented by first changing

        strcpy(buffer, "+proj=pipeline +step +init=");

to

    strcpy(buffer, "+proj=pipeline +towgs84=0,0,0 +step +init=");

Then going on with the setup, but adding some 15 lines of code, checking the case of identical Helmert parameters (instantiate, check), and then finally:

    if (helmerts_not_identical)
        return P;
    proj_destroy (P);
    t = strstr (buffer, "towgs84");
    strcpy (t, "tonight"); /* or xxxxxxx or whatever fits */
    P = proj_create(ctx, buffer);
    return P;

Simpler and more localized than solving the general problem in `pj_pipeline`

So, by and large, see the handheld introduction of towgs84=0,0,0 in the `*.gie` files as a temporary exercise, helping pinpoint some cases where we can do *somewhat* mechanically introduced optimizations in a comming (version 6-ish) early version of a *late binding* EPSG implementation.

Reply:
Okay, that works for me. Do you want to make that change to `proj_create_crs_to_crs`? It fits nicely within this PR.

Reply:
Sure, but I'm offline until Monday, so not until then.